In [12]:
import pandas as pd

data_path = "D:/studium/analyse/data_prep/output/data_corpus.csv"
df = pd.read_csv(data_path, usecols=["id", "text_preproc1"])

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123667 entries, 0 to 123666
Data columns (total 2 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             123667 non-null  object
 1   text_preproc1  123667 non-null  object
dtypes: object(2)
memory usage: 1.9+ MB


In [1]:
from bertopic import BERTopic
topic_model = BERTopic.load("output/bertopic_model_v1")

In [2]:
topics, probabilities = topic_model.topics_, topic_model.probabilities_

In [3]:
topic_model.get_topic_info(-1).Count[0]

17982

#### 1
```
Topic -1 Count: 1991
umap_model = UMAP(n_neighbors=10, min_dist=0.0, n_components=8, random_state=42)
hdbscan_model = hdbscan.HDBSCAN(min_cluster_size=60, min_samples=2, cluster_selection_epsilon=0.1)
topic_model = BERTopic(hdbscan_model=hdbscan_model, umap_model=umap_model, embedding_model=model)
```

In [4]:
freq = topic_model.get_topic_info();

0	-1	23358	-1_der_die_und_object
1	0	2304	0_exoskeleton_limb_rehabilitation_waist
2	1	1400	1_surgical_instrument_surgery_patient
3	2	970	2_welding_seam_gun_weld
4	3	602	3_pipeline_pipe_dredging_oil
5	4	508	4_charging_pile_recharging_charged
6	5	429	5_planning_path_global_algorithm
7	6	398	6_image_images_captured_camera

0	-1	17982	-1_der_und_die_eine
1	0	3375	0_underwater_cleaning_water_dust
2	1	3332	1_surgical_instrument_effector_surgery
3	2	2580	2_cloud_path_map_planning
4	3	2293	3_exoskeleton_limb_rehabilitation_waist
5	4	1219	4_map_apparatus_program_trajectory
6	5	1045	5_chassis_carrying_tray_goods
7	6	997	6_welding_seam_gun_weld

In [5]:
freq[['Topic', 'Count', 'Name']].head(30)

Topic  Count                                         Name
0      -1  17982                          -1_system_for_an_in
1       0  19473                0_cleaning_body_robot_welding
2       1   3661       1_surgical_instrument_robotic_effector
3       2   2293      2_exoskeleton_limb_rehabilitation_joint
4       3   1014         3_substrate_sorting_transfer_chamber
5       4    619              4_rpa_voice_automation_dialogue
6       5    247                5_aerial_wind_unmanned_flying
7       6    239             6_food_cooking_kitchen_processor
8       7    159                    7_cup_glass_suction_fluid
9       8    143                    8_agv_gait_biped_carrying
10      9    128    9_catheter_interventional_guidewire_guide
11     10    123                10_printing_print_printer_ink
12     11    108        11_brake_braking_band_electromagnetic
13     12     99              12_tactile_sensor_layer_sensing
14     13     81            13_fruit_picking_fruits_vegetable
15     14     81        14_neural_network_training_neuronalen
16     15     66              15_gear_planetary_wave_harmonic
17     16     54            16_snakelike_snake_shaped_section
18     17     53  17_der_einer_werkstückträgervorrichtung_des
19     18     50      18_vehicle_passenger_autonomous_reality

In [627]:
topic_model.get_topic(5)

[('aerial', 0.5583696692712207),
 ('wind', 0.4762240339759627),
 ('unmanned', 0.4664865895009757),
 ('flying', 0.4525764447657768),
 ('turbine', 0.41122533731574307),
 ('wing', 0.3922356590535333),
 ('blade', 0.3767734351352035),
 ('flight', 0.34517307599310126),
 ('flapping', 0.34396367955885493),
 ('air', 0.34376343103044876)]

In [6]:
fig = topic_model.visualize_topics(); fig

In [629]:
from scipy.cluster import hierarchy as sch

# Hierarchical topics
linkage_function = lambda x: sch.linkage(x, 'single', optimal_ordering=True)
hierarchical_topics = topic_model.hierarchical_topics(texts, linkage_function=linkage_function)

100%|██████████| 18/18 [00:00<00:00, 237.40it/s]


In [630]:
topic_model.visualize_hierarchy(hierarchical_topics=hierarchical_topics)

In [19]:
def print_hierarchical_topics(topic_string):
    lines = topic_string.split('\n')
    for line in lines:
        print(line)

topic_tree = topic_model.get_topic_tree(hierarchical_topics)

print_hierarchical_topics(topic_tree)

NameError: name 'hierarchical_topics' is not defined

In [632]:
# topics_to_merge = [[1, 2],
#                    [3, 4]]
# topic_model.merge_topics(docs, topics_to_merge)

In [7]:
topic_model.visualize_heatmap()

In [8]:
topic_model.visualize_term_rank()

In [637]:
topic_model.visualize_barchart()

In [638]:
# topic_model.visualize_documents(texts)

In [639]:
# topic_model.visualize_distribution(probabilities[200], min_probability=0.015)

In [640]:
topic_model.reduce_topics(texts, nr_topics=20)

2024-05-28 16:28:13,258 - BERTopic - Topic reduction - Reducing number of topics
2024-05-28 16:28:13,259 - BERTopic - Topic reduction - Reduced number of topics from 20 to 20


In [11]:
topic_model.generate_topic_labels()

['-1_system_for_an',
 '0_cleaning_body_robot',
 '1_surgical_instrument_robotic',
 '2_exoskeleton_limb_rehabilitation',
 '3_substrate_sorting_transfer',
 '4_rpa_voice_automation',
 '5_aerial_wind_unmanned',
 '6_food_cooking_kitchen',
 '7_cup_glass_suction',
 '8_agv_gait_biped',
 '9_catheter_interventional_guidewire',
 '10_printing_print_printer',
 '11_brake_braking_band',
 '12_tactile_sensor_layer',
 '13_fruit_picking_fruits',
 '14_neural_network_training',
 '15_gear_planetary_wave',
 '16_snakelike_snake_shaped',
 '17_der_einer_werkstückträgervorrichtung',
 '18_vehicle_passenger_autonomous']

In [18]:
topic_model.get_representative_docs()

{-1: ['Robot. The invention relates to the technical field of automatic equipment, in particular to a robot which comprises a head assembly, a first display assembly and a first signal collecting assembly are arranged on the head assembly, a second display assembly is arranged on the side face, facing the outer side, of a robot body, and the first signal collecting assembly is used for collecting external to be input information, the head assembly can move under the control of a control instruction and or to be input information, and the first display assembly and the second display assembly can perform interaction and information display under the control of the control instruction and or the to be input information. According to the robot provided by the invention, the robot is provided with the head assembly, the display assembly and the signal acquisition assembly which have the interaction function, so that the diversity of the interaction process and the interaction mode of the r

In [15]:
df['topic'] = topic_model.reduce_outliers(df['text_preproc1'], topics)

100%|██████████| 18/18 [00:24<00:00,  1.38s/it]


ValueError: Length of values (46673) does not match length of index (123667)